modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java

```diff
@@ -600,8 +589,11 @@ protected Assignment getAssignment(User.ID userID, Application.Name applicationN
         }
 
         if (currentTime > experiment.getEndTime().getTime()) {
+            // the experiment moves to the PAUSED state if the endtime is reached
+            experiment.setState(Experiment.State.PAUSED);
+            repository.updateExperiment(experiment);
```

Comment:
@AndreaSuckro This code is potentially going to update the state of experiment to be PAUSED and so send multiple emails, because of two reasons #1 Race condition in the multiple threads and #2 Metadata cache refreshes at the frequency of 5mins so assignment flow will not see the state change and so keep coming here... Please refer to the isAssignmentValidForRapidExperiment() implementation for the reference...